File: LeetCode75_LevelTwo/Partition_Equal_Subset_Sum/Partition_Equal_Subset_Sum.py
# 416. Partition Equal Subset Sum Python

class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        count = sum(nums)
        if count % 2 != 0: return False
        subset_sum = count // 2

        dp = [False] * (subset_sum + 1)
        dp[0] = True

        for curr in nums:
            for j in range(subset_sum, curr - 1, -1):
                dp[j] = dp[j] or dp[j-curr]

        return dp[subset_sum]

File: LeetCode75_LevelTwo/Partition_Equal_Subset_Sum/Partition_Equal_Subset_Sum.cc
// 416. Partition Equal Subset Sum C++
#include <vector>
using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (auto &x: nums) sum += x;
        if (sum % 2 != 0) return false;
        int subset_sum = sum / 2;
        
        vector<bool> v(subset_sum + 1, false);
        v[0] = true;
        for (auto &x: nums) {
            for (int i = subset_sum; i >= x; i--){
                if (v[i] || v[i-x]) v[i] = true;
                else v[i] = false;
            }
        }
        return v[subset_sum];
    }
};
